### PR TITLE
Add CONTRIBUTING file

### PR DESCRIPTION
**Pull request by sivaraam**

### CODE REVIEW

**Comment by domdomegg** [comment]
Brilliant content, but thought it might be better suited on the wiki as there's already a lot there? I agree it'd definitely be worth having a CONTRIBUTORS.md file with a link to the wiki.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
+Thanks for considering to contribute to this project! A few guidelines for
+people who want to contribute their code to this software are documented in
+this document. If you're not sure where to start head on to [this wiki page](https://github.com/commons-app/apps-android-commons/wiki/Volunteers-welcome!).
+
+### Make separate commits for logically separate changes.
+
+Always make a commit with complete commit message for a logically
+separate change. It is a good discipline. For details about how to
+write a good commit message, see thei "Describe your changes" well section.
```
Comment (nicolas-raoul) [inline CONTRIBUTING.md]: Looks like the sentence is not really finished:
thei?
well?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+sure that the entire test suite passes.
+
+If you have an account at GitHub (and you can get one for free to work
+on open source projects), you can use their Travis CI integration to
+test your changes.
+
+
+### Update the Wiki page
+
+Do not forget to update the [Wiki page](https://github.com/commons-app/apps-android-commons/wiki)
```
Comment (nicolas-raoul) [inline CONTRIBUTING.md]: Maybe plural "pages" as there are many?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+on GitHub to describe the updated behavior and make sure that the resulting
```
Comment (nicolas-raoul) [inline CONTRIBUTING.md]: Shouldn't you remove the word "resulting"?
Any page can become stale due to a commit.

**Comment by misaochan** [comment]
Thanks for submitting this!

My main concern is that I really doubt the vast majority of contributors will read the entirety of that file before they start contributing. I wonder if we could have a "cliffs notes" version of this (a few bullet points summarizing everything) in the [Volunteers Welcome](https://github.com/commons-app/apps-android-commons/wiki/Volunteers-welcome%21) wiki page?

**Comment by janpio** [comment]
As we try to have a somewhat useful documentation in the wiki, I agree that most of the content of this change could actually live there. 

But having the file with a short introduction and links to the relevant Wiki pages actually makes total sense.

File: CONTRIBUTING.md (hunk 3 above)
Comment (sivaraam) [reply CONTRIBUTING.md]: I thought it made sense as I just wanted them to be sure the pages don't become stale as a consequence of *their change*.

**Comment by sivaraam** [comment]
> My main concern is that I really doubt the vast majority of contributors will read the entirety of that file before they start contributing, especially as there's already a lot of starter information on the wiki that they are expected to read.

I don't think they would either that's why I named it "CONTRIBUTING.md" so that GitHub would point them to read it when they make a PR.

>  I wonder if we could have a "cliffs notes" version of this (a few bullet points summarizing the main points) in the wiki, maybe in the Volunteers Welcome page?

Yeah, I'm ok with a cliff notes version of this on the Wiki page.

**Comment by sivaraam** [comment]
> As we try to have a somewhat useful documentation in the wiki, I agree that most of the content of this change could actually live there.

> But having the file with a short introduction and links to the relevant Wiki pages actually makes total sense 

As stated before I like to go the other way round because I think it's better to *draw a line* as to what should be on the Wiki page and what should be present along with the source itself. I like to *draw a line* because keeping everything in the Wiki could just clutter it up.

Currently there seems to be a lot of technical documentation related to the app on the wiki page so I guess we should use it for that. I think we should keep things like CONTRIBUTING.md as text files along with the source.

That's just my opinion. Let's see what the consensus is.

**Comment by sivaraam** [comment]
> I think it's better to draw a line as to what should be on the Wiki page and what should be present along with the source itself.

To use more precise terms, I think we should keep *Documentation* on the wiki pages and *Guidelines*
with the source itself.

File: CONTRIBUTING.md (hunk 2 above)
Comment (sivaraam) [reply CONTRIBUTING.md]: Yes, I made the change will push it after I get a reply for the following comment.

**Comment by whym** [comment]
Is it an adaptation of another project's CONTRIBUTING file? If so, I would be a bit concerned about the licensing status of the original file. (Is it under the Apache license or a compatible license?) It appears like at least some part of this is [shared with others](https://www.google.co.jp/search?source=hp&q=%22If+your+description+starts+to+get+too+long%2C+that%27s+a+sign+that+you%22).

**Comment by sivaraam** [comment]
> Is it an adaptation of another project's CONTRIBUTING file? 

I guess I stated that in the [initial commit message](https://github.com/commons-app/apps-android-commons/pull/879/commits/1802e290b06fd75486430fd42f3d4388970f8cd3) of this PR clearly. Quoting the relevant portion,

> This initial version of guideline is based on the git project's guideline for submitting patches (Documentation/SubmittingPatches).

> If so, I would be a bit concerned about the licensing status of the original file. (Is it under the Apache license or a compatible license?) 

Licensing, I didn't think of that. Lets' see ... I guess it's licensed under GPL v2 as noted in that project's [COPYING](https://github.com/git/git/blob/master/COPYING) file. It doesn't seem to be part of the project that falls under LGPL as noted in their [LGPL-2.1 file](https://github.com/git/git/blob/master/LGPL-2.1). Digging a little I found that Apache licenses **[are not** compatible](https://en.wikipedia.org/wiki/Apache_License#GPL_compatibility) with GPL v2 [link 1]. Though, I'm not good with licensing issues I guess this PR has to be abandoned for this sole reason. Correct me if I'm wrong.

**Comment by whym** [comment]
Thanks for doing the research @sivaraam. You are probably right in that it's GPL and I have the same impression that using a GPL-licensed work in an Apache-licensed work is not going to work well.

I think we can keep it in a wiki page, stating that the page is under GPL, if we want. Alternatively, it might be possible to reword the entire document, retaining the key points but not retaining exact phrasings.

**Comment by sivaraam** [comment]
> Thanks for doing the research @sivaraam. You are probably right in that it's GPL and I have the same impression that using a GPL-licensed work in an Apache-licensed work is not going to work well.

Wait, I missed something. I told that Apache licenses are incompatible with GPL v2 but does that also imply that GPL v2 is **incompatible with** Apache licenses? IOW, can we create a document licensed under Apache which has part of it's content taken from another document which is licensed under GPL v2 ? *Note:* I'm not so good at licensing issues.

BTW, In case the answer to my latter question in the above paragraph is **No**. Is it enough if we get permission from the `git` people to use content from their document? I don't think they would say "No".

**Comment by nicolas-raoul** [comment]
Unfortunately, you can take take something that is GPL and put it inside that is Apache-licensed.

That's because Apache requires reusers to attribute. By putting GPL content in an Apache whole, you would be taking from reusers the liberty to reuse without attribution.

**Comment by sivaraam** [comment]
@nicolas-raoul You meant * ... you **can't** take ..* right?

**Comment by nicolas-raoul** [comment]
@sivaraam : Yes thanks fixed! :-)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+Thanks for considering to contribute to this project! A few guidelines for
+people who want to contribute their code to this software are documented at
+this project's Wiki. If you're not sure where to start head on to [this wiki page](https://github.com/commons-app/apps-android-commons/wiki/Volunteers-welcome!).
+
+Here's a jist of the guidelines,
```
Comment (nicolas-raoul) [inline CONTRIBUTING.md]: list? :-)

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+# Make separate commits for logically separate changes
+
+# Describe your changes well in the commit message
+
+The first line of the commit message should be a short description of what has
+changed. It is also good to prefix the first line with "area: " where the "area"
+is a filename or identifier for the general area of the code being modified.
+The body should provide a meaningful commit message.
+
+# Write tests for your code
```
Comment (nicolas-raoul) [inline CONTRIBUTING.md]: How about adding "if possible"?
Sometimes a UI change can not really be tested via unit tests.

**Comment by neslihanturan** [comment]
@sivaraam if requested changes are done, can you solve conflicts so that I can merge this.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1 +1,19 @@
-Please see our guidelines in the wiki: https://github.com/commons-app/apps-android-commons/wiki/Volunteers-welcome%21
+Thanks for considering to contribute to this project! A few guidelines for
+people who want to contribute their code to this software are documented in
+[this project's Wiki](https://github.com/commons-app/apps-android-commons/wiki/Contributing-Guidelines).
+If you're not sure where to start head on to [this wiki page](https://github.com/commons-app/apps-android-commons/wiki/Volunteers-welcome!).
+
+Here's a jist of the guidelines,
```
Comment (maskaravivek) [inline CONTRIBUTING.md]: Use `gist`

Ref: https://writingexplained.org/jist-or-gist-difference

Comment (sivaraam) [reply CONTRIBUTING.md]: Nice catch. Thanks for the correction!

**Comment by ujjwalagrawal17** [comment]
@maskaravivek I think we can merge this PR. It does not have merge conflicts now.

**Comment by neslihanturan** [comment]
As far I as follow all reviews are done, so merging this. Thanks @sivaraam !

**Pull request merged**